worker/src/sidechain_storage/storage.rs

```diff
@@ -0,0 +1,887 @@
+						current_shard, current_block_nr
+					);
+					continue
+				}
+			} else {
+				self.shards.push(current_shard);
+				new_shard = true;
+			}
+			// Block hash -> Signed Block
+			let current_block_hash = signed_block.hash();
+			batch.put(&current_block_hash.encode(), &signed_block.encode().as_slice());
+			// (Shard, Block number) -> Blockhash (for block pruning)
+			batch.put(
+				&(current_shard, current_block_nr).encode().as_slice(),
+				&current_block_hash.encode(),
+			);
+			// (last_block_key, shard) -> (Blockhash, BlockNr) current blockchain state
+			let current_last_block =
+				LastSidechainBlock { hash: current_block_hash, number: current_block_nr };
+			self.last_blocks.insert(current_shard, current_last_block.clone());
+			batch.put((LAST_BLOCK_KEY, current_shard).encode(), current_last_block.encode());
+		}
+		// update stored_shards_key -> vec<shard> only if a new shard was included
+		if new_shard {
+			batch.put(STORED_SHARDS_KEY.encode(), self.shards.encode());
+		}
+		self.db.write(batch).map_err(Error::OperationalError)
+	}
+
+	/// purges a shard and its block from the db storage
+	pub fn purge_shard(&mut self, shard: &ShardIdentifier) -> Result<()> {
+		if self.shards.contains(shard) {
+			// get last block of shard
+			let mut last_block = match self.last_blocks.get(shard) {
+				Some(last_block) => last_block.clone(),
+				None => return Err(Error::LastBlockNotFound(*shard)),
+			};
+			// remove last block from db storage
+			// Blockhash -> Signed Block
+			self.db.delete(last_block.hash.encode()).map_err(Error::OperationalError)?;
+			// (Shard , Block number) -> Blockhash
+			self.db
+				.delete((shard, last_block.number).encode())
+				.map_err(Error::OperationalError)?;
+			// (LAST_BLOCK_KEY, Shard) -> LastSidechainBlock
+			self.db
+				.delete((LAST_BLOCK_KEY, shard).encode())
+				.map_err(Error::OperationalError)?;
+			self.last_blocks.remove(&shard); // delete from local memory
+
+			// Remove all blocks from db
+			while let Some(previous_block) = self.get_previous_block(shard, last_block.number) {
+				last_block = previous_block;
+				// Blockhash -> Signed Block
+				self.db.delete(last_block.hash.encode()).map_err(Error::OperationalError)?;
+				// (Shard, Block number) -> Blockhash
+				self.db
+					.delete((shard, last_block.number).encode())
+					.map_err(Error::OperationalError)?;
+			}
+			// remove shard from list
+			// STORED_SHARDS_KEY -> Vec<(Shard)>
+			self.shards.retain(|&x| x != *shard);
+			self.db
+				.put(STORED_SHARDS_KEY.encode(), self.shards.encode())
+				.map_err(Error::OperationalError)?
+		}
+		Ok(())
+	}
+
+	/// purges a shard and its block from the db storage
+	pub fn prune_shard_from_block_number(
+		&mut self,
+		shard: &ShardIdentifier,
+		block_number: BlockNumber,
+	) -> Result<()> {
+		if self.shards.contains(&shard) {
+			// get last block of shard
+			let last_block = match self.last_blocks.get(shard) {
```

Comment:
Here the clone should not be needed either. In such a case, I like to do then:

```
self.last_blocks.get().ok_ok(Error::LastBlockNotFound(*shard)?;
```

Reply:
Sure, that looks better. Implemented in commit https://github.com/integritee-network/worker/pull/380/commits/3441fc95581f7ecd7e6ab7c4d8d165db2d0462f2. Atleast partially ;)